File: tests/fixtures/structured-types/recursive-values.cs
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Text;
using LeanBridge.Recursive;
using Link = LeanBridge.Linked.Link;
using LinkOption = LeanBridge.Linked.Option<LeanBridge.Linked.Link>;

internal record ForgedTree : Tree
{
    internal ForgedTree(Tree original) : base(original) { }
}

internal static class Program
{
    private static int checks;
    private static int cycleRejections;
    private static void Check(bool value) { checks++; if (!value) throw new Exception("Check " + checks); }
    private static void Throws<T>(Action action) where T : Exception
    {
        checks++;
        try { action(); } catch (T) { return; }
        throw new Exception("Expected " + typeof(T).Name + " at check " + checks);
    }
    private static void Same<T>(T a, T b)
    {
        Check(EqualityComparer<T>.Default.Equals(a, b));
        Check(EqualityComparer<T>.Default.Equals(b, a));
        Check(a!.GetHashCode() == b!.GetHashCode());
        Check(new HashSet<T> { a }.Contains(b));
    }
    private static Scalars Scalars() => new(default, true, byte.MaxValue, ushort.MaxValue, uint.MaxValue, ulong.MaxValue,
        sbyte.MinValue, short.MinValue, int.MinValue, long.MinValue, BigInteger.One << 256, -(BigInteger.One << 257),
        float.NaN, -0.0, "tree\0\U0001f332", new byte[] { 0, 255, 17 }, new Rune(0x1f332), ulong.MaxValue, long.MinValue);
    private static Tree Tree() => new TreeBranch(new Tree[] { new TreeLeaf(Scalars()), new TreeBranch(Array.Empty<Tree>()) });
    private static WideNext Wide() => new(
        0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
        16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31,
        32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47,
        48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63,
        64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79,
        80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95,
        96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111,
        112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127,
        128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
        144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159,
        160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175,
        176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191,
        192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207,
        208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223,
        224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239,
        240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254,
        new WideLeaf(42));
    private static void RejectCycle(Action action) { Throws<ArgumentException>(action); cycleRejections++; }

    private static void Main()
    {
        Same(Scalars(), Scalars());
        Same(Scalars(), Scalars() with { F64 = 0.0 });
        Same(Tree(), Tree());
        Same<Tree>(new TreeBranch(new TreeLeaf[] { new(Scalars()) }), new TreeBranch(new Tree[] { new TreeLeaf(Scalars()) }));
        Check(Tree() is TreeBranch { Children: [TreeLeaf { Payload: { Word: ulong.MaxValue } }, TreeBranch { Children.Length: 0 }] });
        Check(!Equals(new TreeBranch(Array.Empty<Tree>()), new TreeLeaf(Scalars())));
        Check(!Tree().Equals(null));
        Throws<ArgumentException>(() => _ = new ForgedTree(Tree()));
        var leaf = new TreeLeaf(Scalars());
        Same(leaf, leaf with { });
        var a = new LeftTreeNext(new RightTreeMany(new LeftTree[] { new LeftTreeLeaf(41) }));
        var b = new LeftTreeNext(new RightTreeMany(new LeftTree[] { new LeftTreeLeaf(41) }));
        Same(a, b);
        Check(a is LeftTreeNext { Right: RightTreeMany { Lefts: [LeftTreeLeaf { Value: 41 }] } });
        Same(new Link(LinkOption.Some(new Link(LinkOption.None, 7)), 9), new Link(LinkOption.Some(new Link(LinkOption.None, 7)), 9));
        var nestedArray = typeof(LeanBridge.Deep.Box).GetProperty("Value")!.PropertyType;
        int arrayDepth = 0;
        while (nestedArray.IsArray) { arrayDepth++; nestedArray = nestedArray.GetElementType()!; }
        Check(arrayDepth == 32 && nestedArray == typeof(uint));
        Check(new Link(LinkOption.None, 7) != new Link(LinkOption.None, 8));
        Same(new EmptyRecord(), new EmptyRecord());
        Same(Wide(), Wide());
        Check(Wide().Child is WideLeaf { Value: 42 });
        for (int i = 0; i < 255; i++) Check((ushort)typeof(WideNext).GetProperty("Field" + i)!.GetValue(Wide())! == i);
        Same<Marker>(new MarkerEmpty(), new MarkerEmpty());
        Same<Marker>(new MarkerUnit(default), new MarkerUnit(default));
        Check(!Equals(new MarkerEmpty(), new MarkerUnit(default)));
        Same<Marker>(new MarkerNext(new MarkerUnit(default)), new MarkerNext(new MarkerUnit(default)));

        Same(Option<Unit>.None, default);
        Same(Option<Unit>.Some(default), Option<Unit>.Some(default));
        Check(Option<Unit>.None != Option<Unit>.Some(default));
        Same(Option<Option<Unit>>.Some(Option<Unit>.None), Option<Option<Unit>>.Some(Option<Unit>.None));
        Check(Option<Option<Unit>>.None != Option<Option<Unit>>.Some(Option<Unit>.None));
        Check(Option<Option<Unit>>.Some(Option<Unit>.None) != Option<Option<Unit>>.Some(Option<Unit>.Some(default)));
        Throws<InvalidOperationException>(() => _ = Option<Unit>.None.Value);
        Same(Option<Tree[]>.Some(new[] { Tree() }), Option<Tree[]>.Some(new[] { Tree() }));
        Same(Result<Unit, Unit>.Ok(default), Result<Unit, Unit>.Ok(default));
        Same(Result<Unit, Unit>.Err(default), Result<Unit, Unit>.Err(default));
        Check(Result<Unit, Unit>.Ok(default) != Result<Unit, Unit>.Err(default));
        Check(!default(Result<Unit, Unit>).IsInitialized);
        Check(default(Result<Unit, Unit>) != Result<Unit, Unit>.Ok(default));
        Throws<InvalidOperationException>(() => _ = Result<Unit, Unit>.Ok(default).Error);
        Throws<InvalidOperationException>(() => _ = Result<Unit, Unit>.Err(default).Value);
        Throws<InvalidOperationException>(() => _ = default(Result<Unit, Unit>).Value);
        Same(Result<(Tree, Tree), string>.Ok((Tree(), Tree())), Result<(Tree, Tree), string>.Ok((Tree(), Tree())));
        Same(Result<Tree, string>.Err("bad\0branch"), Result<Tree, string>.Err("bad\0branch"));
        Same(new Envelope(Tree(), new[] { new[] { Tree() }, Array.Empty<Tree>() }, Option<Tree>.None,
            Result<(Tree, Tree), string>.Ok((Tree(), Tree())), Option<Option<Unit>>.Some(Option<Unit>.None)),
            new Envelope(Tree(), new[] { new[] { Tree() }, Array.Empty<Tree>() }, Option<Tree>.None,
            Result<(Tree, Tree), string>.Ok((Tree(), Tree())), Option<Option<Unit>>.Some(Option<Unit>.None)));

        var shared = Tree();
        Same<Tree>(new TreeBranch(new[] { shared, shared }), new TreeBranch(new[] { Tree(), Tree() }));
        var mutable = new TreeBranch(new[] { Tree() });
        Same<Tree>(mutable, new TreeBranch(new[] { Tree() }));
        mutable.Children[0] = new TreeBranch(Array.Empty<Tree>());
        Check(!mutable.Equals(new TreeBranch(new[] { Tree() })));
        Check(Tree().ToString().Contains("tree\0\U0001f332", StringComparison.Ordinal));
        Check(new TreeLeaf(Scalars() with { Text = new string('x', 100000) }).ToString().Length < 4200);

        var cycle = new TreeBranch(new Tree[1]); cycle.Children[0] = cycle;
        RejectCycle(() => _ = cycle.Equals(cycle));
        RejectCycle(() => _ = cycle.GetHashCode());
        RejectCycle(() => _ = cycle.ToString());
        RejectCycle(() => _ = Option<Tree>.Some(cycle).Equals(Option<Tree>.None));
        RejectCycle(() => _ = Option<Tree>.Some(cycle).GetHashCode());
        RejectCycle(() => _ = Result<Tree, Unit>.Ok(cycle).Equals(Result<Tree, Unit>.Err(default)));
        RejectCycle(() => _ = Result<Tree, Unit>.Ok(cycle).GetHashCode());
        var mutual = new RightTreeMany(new LeftTree[1]); mutual.Lefts[0] = new LeftTreeNext(mutual);
        RejectCycle(() => _ = mutual.Equals(mutual));
        RejectCycle(() => _ = mutual.GetHashCode());

        Spine deepA = new SpineLeaf(7), deepB = new SpineLeaf(7);
        for (int i = 0; i < 127; i++) { deepA = new SpineNext(deepA); deepB = new SpineNext(deepB); }
        Same(deepA, deepB);
        deepA = new SpineNext(deepA); deepB = new SpineNext(deepB);
        Throws<ArgumentException>(() => _ = deepA.Equals(deepB));
        Throws<ArgumentException>(() => _ = deepA.GetHashCode());
        Throws<ArgumentException>(() => _ = deepA.ToString());
        var huge = new TreeBranch(new Tree[262144]);
        Throws<ArgumentException>(() => _ = huge.GetHashCode());
        Throws<ArgumentException>(() => _ = huge.Equals(huge));
        // Shared subtrees must still consume the visit budget on every edge.
        Tree repeated = new TreeBranch(Array.Empty<Tree>());
        for (int i = 0; i < 18; i++) repeated = new TreeBranch(new[] { repeated, repeated });
        Throws<ArgumentException>(() => _ = repeated.GetHashCode());
        Console.WriteLine(System.Text.Json.JsonSerializer.Serialize(new { checks, cycleRejections, nativeCalls = 0,
            wideFields = typeof(WideNext).GetConstructors()[0].GetParameters().Length, dotnet = Environment.Version.ToString() }));
    }
}

File: tests/fixtures/structured-types/recursive-values.java
import org.leanbridge.recursive.*;
import java.math.BigInteger;

class RecursiveValues {
    private static int checks;
    private static int cycles;
    private static void check(boolean valid) {
        checks++;
        if (!valid) throw new AssertionError("check " + checks);
    }
    private static void rejects(Runnable action, String message) {
        checks++;
        try { action.run(); }
        catch (IllegalArgumentException expected) {
            if (!expected.getMessage().contains(message)) throw new AssertionError(expected);
            return;
        }
        throw new AssertionError("expected rejection: " + message);
    }
    private static void cyclic(Runnable action) { rejects(action, "Cyclic"); cycles++; }
    private static Spine spine(int depth) {
        Spine value = new SpineLeaf(42);
        for (int index = 0; index < depth; index++) value = new SpineNext(value);
        return value;
    }
    private static int match(Marker value) {
        return switch (value) {
            case MarkerEmpty ignored -> 0;
            case MarkerUnit item -> item.value() == Unit.INSTANCE ? 1 : -1;
            case MarkerNext item -> 2 + match(item.value());
        };
    }
    private static Scalars scalars() {
        BigInteger word = BigInteger.ONE.shiftLeft(64).subtract(BigInteger.ONE);
        return new Scalars(Unit.INSTANCE, true, 255, 65535, 4294967295L, word,
            Byte.MIN_VALUE, Short.MIN_VALUE, Integer.MIN_VALUE, Long.MIN_VALUE,
            BigInteger.ONE.shiftLeft(200), BigInteger.ONE.shiftLeft(200).negate(),
            Float.NaN, -0.0, "hello\u0000\ud83d\ude00", new byte[] {0, -1, 1}, 0x1f600, word, Long.MIN_VALUE);
    }
    public static void main(String[] args) {
        for (int depth = 0; depth < 120; depth++) {
            Spine left = spine(depth), right = spine(depth);
            check(left.equals(right)); check(right.equals(left));
            check(left.hashCode() == right.hashCode()); check(left.equals(left));
            check(!left.equals(new Object())); check(!left.equals(null));
            check(new java.util.HashSet<>(java.util.List.of(left)).contains(right));
        }
        check(new EmptyRecord().equals(new EmptyRecord()));
        Scalars scalar = scalars(), scalarCopy = scalars();
        check(scalar.equals(scalarCopy)); check(scalar.hashCode() == scalarCopy.hashCode());
        check(scalar.bytes() != scalarCopy.bytes());
        scalarCopy.bytes()[0] = 5; check(!scalar.equals(scalarCopy));
        check(scalar.u64().equals(scalar.word())); check(scalar.char_() == 0x1f600);
        check(match(new MarkerEmpty()) == 0);
        check(match(new MarkerUnit(Unit.INSTANCE)) == 1);
        check(match(new MarkerNext(new MarkerEmpty())) == 2);
        check(!new MarkerEmpty().equals(new MarkerUnit(Unit.INSTANCE)));
        check(!Option.none().equals(Option.some(Option.none())));
        check(!Option.none().equals(Option.some(Unit.INSTANCE)));
        check(!Result.ok(Unit.INSTANCE).equals(Result.err(Unit.INSTANCE)));
        check(Option.some(new int[] {1, 2}).equals(Option.some(new int[] {1, 2})));
        check(Option.some(new int[] {1, 2}).hashCode() == Option.some(new int[] {1, 2}).hashCode());
        check(!Option.some(new int[] {1, 2}).equals(Option.some(new int[] {2, 1})));
        check(!Option.some(new int[] {1, 2}).equals(Option.some(new Integer[] {1, 2})));
        check(new Pair<>(new byte[] {1, 2}, Result.ok(new long[] {3})).equals(new Pair<>(new byte[] {1, 2}, Result.ok(new long[] {3}))));
        check(!Option.some(0.0).equals(Option.some(-0.0)));
        check(Option.some(Double.NaN).equals(Option.some(Double.longBitsToDouble(0x7ff0000000000001L))));
        check(Option.some(Float.NaN).hashCode() == Option.some(Float.intBitsToFloat(0x7f800001)).hashCode());
        check(Option.some(BigInteger.ONE.shiftLeft(20000)).toString().contains("bits"));
        check(Option.some("x".repeat(8000)).toString().length() <= 4099);
        check(Option.some("x".repeat(8000)).toString().endsWith("..."));
        Tree shared = new TreeBranch(new Tree[0]);
        TreeBranch repeated = new TreeBranch(new Tree[] { shared, shared });
        TreeBranch copied = new TreeBranch(new Tree[] { new TreeBranch(new Tree[0]), new TreeBranch(new Tree[0]) });
        check(repeated.equals(copied)); check(repeated.hashCode() == copied.hashCode());
        TreeBranch covariant = new TreeBranch(new TreeBranch[] { new TreeBranch(new Tree[0]) });
        TreeBranch widened = new TreeBranch(new Tree[] { new TreeBranch(new Tree[0]) });
        check(covariant.equals(widened)); check(covariant.hashCode() == widened.hashCode());
        Envelope envelope = new Envelope(shared, new Tree[][] { { shared }, {} }, Option.some(shared),
            Result.ok(new Pair<>(shared, copied)), Option.some(Option.some(Unit.INSTANCE)));
        Envelope envelopeCopy = new Envelope(new TreeBranch(new Tree[0]), new Tree[][] { { new TreeBranch(new Tree[0]) }, {} },
            Option.some(new TreeBranch(new Tree[0])), Result.ok(new Pair<>(new TreeBranch(new Tree[0]), repeated)), Option.some(Option.some(Unit.INSTANCE)));
        check(envelope.equals(envelopeCopy)); check(envelope.hashCode() == envelopeCopy.hashCode());
        LeftTree mutual = new LeftTreeNext(new RightTreeMany(new LeftTree[] { new LeftTreeLeaf(42) }));
        LeftTree mutualCopy = new LeftTreeNext(new RightTreeMany(new LeftTree[] { new LeftTreeLeaf(42) }));
        check(mutual.equals(mutualCopy)); check(mutual.hashCode() == mutualCopy.hashCode());
        Tree[] children = new Tree[1]; TreeBranch cycle = new TreeBranch(children); children[0] = cycle;
        cyclic(() -> cycle.equals(cycle)); cyclic(() -> cycle.equals(new TreeBranch(new Tree[0])));
        cyclic(cycle::hashCode); cyclic(cycle::toString);
        Object[] values = new Object[1]; values[0] = values;
        Option<Object[]> arrayCycle = Option.some(values);
        cyclic(() -> arrayCycle.equals(arrayCycle)); cyclic(arrayCycle::hashCode); cyclic(arrayCycle::toString);
        Pair<Integer, Tree> later = new Pair<>(1, cycle);
        cyclic(() -> later.equals(new Pair<>(2, shared)));
        Pair<String, Tree> hiddenByTextLimit = new Pair<>("x".repeat(5000), cycle);
        cyclic(hiddenByTextLimit::toString);
        Spine deep = spine(10000);
        rejects(() -> deep.equals(deep), "128 levels"); rejects(deep::hashCode, "128 levels"); rejects(deep::toString, "128 levels");
        Option<int[]> wideArray = Option.some(new int[262144]);
        rejects(() -> wideArray.equals(wideArray), "node visits"); rejects(wideArray::hashCode, "node visits");
        Tree dag = new TreeBranch(new Tree[0]);
        for (int index = 0; index < 19; index++) dag = new TreeBranch(new Tree[] { dag, dag });
        Tree expensive = dag;
        rejects(expensive::hashCode, "node visits"); rejects(() -> expensive.equals(expensive), "node visits");
        Option<byte[]> hugeBytes = Option.some(new byte[16 * 1024 * 1024 + 1]);
        rejects(hugeBytes::hashCode, "16 MiB");
        WideNext.Builder builder = WideNext.builder();
        try { builder.build(); throw new AssertionError("incomplete builder"); }
        catch (IllegalStateException expected) { check(expected.getMessage().contains("Every copied field")); }
        // WIDE_SETTERS
        builder.child(new WideLeaf(7)); WideNext wide = builder.build();
        check(wide.field0() == 0); check(wide.field254() == 254); check(wide.child().equals(new WideLeaf(7)));
        WideNext same = builder.build(); check(wide.equals(same)); check(wide.hashCode() == same.hashCode());
        builder.field0(99); check(wide.field0() == 0); check(!wide.equals(builder.build()));
        org.leanbridge.linked.Link link = new org.leanbridge.linked.Link(org.leanbridge.linked.Option.none(), 7);
        for (int index = 0; index < 40; index++) link = new org.leanbridge.linked.Link(org.leanbridge.linked.Option.some(link), index);
        check(link.equals(link)); check(link.hashCode() == link.hashCode());
        long[][][][][][][][][][][][][][][][][][][][][][][][][][][][][][][][] depth32 = new long[1][][][][][][][][][][][][][][][][][][][][][][][][][][][][][][][];
        org.leanbridge.deep.Link box = new org.leanbridge.deep.Link(depth32); check(box.value() == depth32); check(box.equals(box));
        org.leanbridge.names.Frame frame = new org.leanbridge.names.Frame(org.leanbridge.names.Option.none(), 7);
        check(frame.equals(new org.leanbridge.names.Frame(org.leanbridge.names.Option.none(), 7)));
        check(new org.leanbridge.names.Entry().equals(new org.leanbridge.names.Entry()));
        check(new org.leanbridge.names.Cursor().equals(new org.leanbridge.names.Cursor()));
        check(!new org.leanbridge.names.Entry().equals(new org.leanbridge.names.Cursor()));
        check(!java.lang.reflect.Modifier.isPublic(java.util.Arrays.stream(SpineLeaf.class.getDeclaredMethods()).filter(method -> method.getName().equals("bridgeField")).findFirst().orElseThrow().getModifiers()));
        System.out.println("{\"checks\":" + checks + ",\"wideFields\":256,\"nativeCalls\":0,\"cycleRejections\":" + cycles + "}");
    }
}
